searchcore/src/tests/proton/attribute/document_field_extractor/document_field_extractor_test.cpp

```diff
@@ -0,0 +1,380 @@
+{
+    auto result = std::make_unique<WeightedSetFieldValue>(weightedSetTypeInt);
+    for (const auto &elem : array) {
+        result->add(IntFieldValue(elem.first), elem.second);
+    }
+    return result;
+}
+
+std::unique_ptr<FieldValue>
+makeStringWeightedSet(const std::vector<std::pair<vespalib::string, int32_t>> &array)
+{
+    auto result = std::make_unique<WeightedSetFieldValue>(weightedSetTypeString);
+    for (const auto &elem : array) {
+        result->add(StringFieldValue(elem.first), elem.second);
+    }
+    return result;
+}
+
+template <typename F1, typename F2>
+void
+checkFieldPathChange(F1 f1, F2 f2, const vespalib::string &path, bool same)
+{
+    FieldPath fieldPath1 = f1.makeFieldPath(path);
+    FieldPath fieldPath2 = f2.makeFieldPath(path);
+    EXPECT_TRUE(!fieldPath1.empty());
+    EXPECT_TRUE(!fieldPath2.empty());
+    EXPECT_TRUE(DocumentFieldExtractor::isSupported(fieldPath1));
+    EXPECT_TRUE(DocumentFieldExtractor::isSupported(fieldPath2));
+    EXPECT_EQUAL(same, DocumentFieldExtractor::isCompatible(fieldPath1, fieldPath2));
+}
+
+}
+
+struct FixtureBase
+{
+    DocumentType   type;
+    const Field    tagField;
+    const Field    nameField;
+
+    FixtureBase(bool byteTag)
+        : type("test"),
+          tagField("tag", 1, byteTag ? *DataType::BYTE : *DataType::INT, true),
+          nameField("name", 2, *DataType::STRING, true)
+    {
+    }
+
+    ~FixtureBase();
+
+    std::unique_ptr<Document>
+    makeDoc()
+    {
+        return std::make_unique<Document>(type, DocumentId("id::test::1"));
+    }
+
+    FieldPath
+    makeFieldPath(const vespalib::string &path)
+    {
+        FieldPath fieldPath;
+        try {
+            type.buildFieldPath(fieldPath, path);
+        } catch (FieldNotFoundException &) {
+            fieldPath = FieldPath();
+        }
+        if (!DocumentFieldExtractor::isSupported(fieldPath)) {
+            fieldPath = FieldPath();
+        }
+        return fieldPath;
+    }
+
+    void
+    assertExtracted(DocumentFieldExtractor &extractor,
+                    const vespalib::string &path,
+                    std::unique_ptr<FieldValue> expected) {
+        FieldPath fieldPath(makeFieldPath(path));
+        std::unique_ptr<FieldValue> fv = extractor.getFieldValue(fieldPath);
+        if (expected) {
+            ASSERT_TRUE(fv);
+            EXPECT_EQUAL(*expected, *fv);
+        } else {
+            EXPECT_TRUE(!fv);
+        }
+    }
+};
+
+FixtureBase::~FixtureBase() = default;
+
+struct SimpleFixture : public FixtureBase
+{
+    SimpleFixture(bool byteTag = false)
+        : FixtureBase(byteTag)
+    {
+        type.addField(tagField);
+        type.addField(nameField);
+    }
+};
+
+TEST_F("require that simple fields give simple values", SimpleFixture)
+{
+    auto doc = f.makeDoc();
+    doc->setValue(f.tagField, IntFieldValue(200));
+    doc->setValue(f.nameField, StringFieldValue("name200b"));
+    DocumentFieldExtractor extractor(*doc);
```

Comment:
Consider keeping extractor inside fixture (setup during makeDoc() and store doc instance as well) to simplify test setup. This will also simplify assertExtracted() signature.